cibyl/sources/zuuld/source.py

```diff
@@ -0,0 +1,38 @@
+"""
+#    Copyright 2022 Red Hat
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+"""
+
+from cibyl.sources.source import Source, speed_index
+from cibyl.sources.zuuld.git_api import ZuulGitHub, ZuulLocal
+
+
+class ZuulD(Source):
+    def __init__(self, name: str, driver: str, **kwargs):
+        super().__init__(name, driver, **kwargs)
+        self.repos = kwargs.get("repos")
+        self.kwargs = kwargs
+        remote = kwargs.get("remote", False)
+        if not remote:
+            self.api = ZuulLocal(self.kwargs)
+        else:
+            self.api = ZuulGitHub(self.kwargs)
+        self.api._validate()
```

Comment:
Calling `_validate` here will end up cloning/pulling the repos when the ZuulD source object is initialised, meaning that we might spend some time setting up something that might not be used (if the user wants to query Jenkins, for example). This can be solved by moving the call inside a setup method, which is the mechanism used in the GitSource. That should ensure that the source will only be setup if it might be needed.

Reply:
Done.

Reply:
Sorry to insist, but while I think that having a validate method could be a good idea, this source should still expose a setup method and call `self.api._validate` from it to avoid cloning multiple repos that might not be used.

Reply:
okay. fixed.